crates/wasmparser/src/binary_reader.rs

```diff
@@ -1553,6 +1553,726 @@ impl<'a> BinaryReader<'a> {
         Ok(BlockType::FuncType(idx as u32))
     }
 
+    /// Reads the next available `Operator` and calls the respective visit method.
+    ///
+    /// # Errors
+    ///
+    /// If `BinaryReader` has less bytes remaining than required to parse
+    /// the `Operator`.
+    #[rustfmt::skip]
+    pub fn visit_operator<T>(&mut self, visitor: &mut T) -> Result<<T as VisitOperator<usize>>::Output>
+    where
+        T: VisitOperator<usize>,
```

Comment:
I made `VisitOperator` generic over this `Input` type which usually is `usize` to neatly integrate the existing validation procedures into the new framework. This was mainly done to keep the code changes as minimal and contained as possible. I also think it is a nice design to have an `Input` and `Output` duality in the trait which makes it very versatile. For `wasmi` I have use cases where `Input = ()` and it is nice to have this `Input` to not have to define yet another entire data structure just to capture the common input as a user.

Reply:
One thing I wanted to add is that I personally would have liked for `VisitOperator<usize>` to actually be something more along the lines of a strongly typed `VisitOperator<Offset>` which also more carries over the purpose of the internal `usize` where `Offset` is of type `struct Offset(usize)`.

Reply:
Generally we could remove the generic parameter of the `VisitOperator` trait together with the `offset: usize` parameter but that would decrease the versatility of the entire trait approach and would cause a huge amount of boilerplate code. By introducing this generic `Input` type we spare us a lot of this boilerplate code and make the trait very versatile for users of the `wasmparser` crate as well as the `wasmparser` crate itself.

Reply:
I'm not sure I understand what you mean by removing this would create a huge amount of boilerplate. My point is that each method is always invoked with a `usize` offset and it's always available from the binary decoder. I'd ideally prefer to not try to design a trait that's super general for all purposes and rather keep it relatively form-fitted for `wasmparser`'s use case.

I don't mind using an `Offset` type but I think that should be in a separate PR because `usize` is used pervasively throughout `wasmparser` as an offset type, so the PR would be quite large to change everything necessary.

Also to clarify I'm not saying that the `usize` argument should be removed from the methods, just the type parameter on the trait because I'm not sure how it would vary and still be callable by this crate.

Reply:
I will see what I can do to remove the generic parameter of the `VisitOperator` trait with non-pervasive changes to `FuncValidator` and `OperatorValidator`.